src/integration_tests/test_with_many_nodes.py

```diff
@@ -0,0 +1,145 @@
+
+        # debugging only, where we keep an instance up between processes
+        self.state, self.statefile = {}, '/tmp/.open-test-instances.txt'
+        project = 'project-with-cluster-integration-tests'
+
+        if self.reuse_existing_stack and os.path.exists(self.statefile):
+            # evidence of a previous instance and we've been told to re-use old instances
+            old_state = json.load(open(self.statefile, 'r'))
+            old_env = old_state.get('environment')
+
+            # test if the old stack still exists ...
+            if old_env and core.describe_stack(project + "--" + old_env, allow_missing=True):
+                self.state = old_state
+                self.environment = old_env
+            else:
+                # nope. old statefile is bogus, delete it
+                os.unlink(self.statefile)
+
+        self.state['environment'] = self.environment # will be saved later
+
+        with settings(abort_on_prompts=True):
+            self.stackname = '%s--%s' % (project, self.environment)
+            self.stacknames.append(self.stackname)
+
+            if self.cleanup:
+                cfn.ensure_destroyed(self.stackname)
+
+            self.context, self.cfn_template, _ = cfngen.generate_stack(project, stackname=self.stackname)
+            self.region = self.context['project']['aws']['region']
+            bootstrap.create_stack(self.stackname)
+
+            # lifecycle.start(self.stackname) # see self.setUp
+
+    @classmethod
+    def tearDownClass(self): # cls, not self
+        super(One, self).tearDownClass()
+        try:
+            if self.reuse_existing_stack:
+                json.dump(self.state, open(self.statefile, 'w'))
+            if self.cleanup:
+                for stackname in self.stacknames:
+                    cfn.ensure_destroyed(stackname)
+            # self.rm_temp_dir()
+            # self.assertFalse(os.path.exists(self.temp_dir), "failed to delete path %r in tearDown" % self.temp_dir)
+        except BaseException:
+            # important, as anything in body will silently fail
+            LOG.exception('uncaught error tearing down test class')
+
+    def setUp(self):
+        lifecycle.start(self.stackname)
+
+    def test_restart_all_started(self):
+        "multiple nodes can be restarted from a running state"
+        history = lifecycle.restart(self.stackname)
+
+        # two nodes rebooting in serial, node 1 first, both running
+        expected_history = [
+            [(1, 'running'), (2, 'running')],
+            [(1, 'stopping'), (2, 'running')],
+            [(1, 'stopped'), (2, 'running')],
+            [(1, 'pending'), (2, 'running')],
+            [(1, 'running'), (2, 'running')],
+
+            # node2, stopped -> running
+            [(1, 'running'), (2, 'stopping')],
+            [(1, 'running'), (2, 'stopped')],
+            [(1, 'running'), (2, 'pending')],
+            [(1, 'running'), (2, 'running')],
+        ]
```

Comment:
the history is a nice way to assert on what has happened

Reply:
when I get frustrated with code and want to tear it all down (which is like, always) I describe exactly what I want and then figure out how to get there. In this case it was pleasantly easy to get to this output.